toolkits/slack/arcade_slack/utils.py

```diff
@@ -0,0 +1,201 @@
+        ConversationType.DIRECT_MESSAGE: ConversationTypeSlackName.IM,
+    }
+    return mapping[conversation_type]
+
+
+def extract_conversation_metadata(conversation: dict) -> dict:
+    conversation_type = get_conversation_type(conversation)
+
+    metadata = {
+        "id": conversation.get("id"),
+        "name": conversation.get("name"),
+        "conversation_type": conversation_type,
+        "is_private": conversation.get("is_private", True),
+        "is_archived": conversation.get("is_archived", False),
+        "is_member": conversation.get("is_member", True),
+        "purpose": conversation.get("purpose", {}).get("value", ""),
+        "num_members": conversation.get("num_members", 0),
+    }
+
+    if conversation_type == ConversationTypeSlackName.IM.value:
+        metadata["num_members"] = 2
+        metadata["user"] = conversation.get("user")
+        metadata["is_user_deleted"] = conversation.get("is_user_deleted")
+    elif conversation_type == ConversationTypeSlackName.MPIM.value:
+        metadata["num_members"] = len(conversation.get("name", "").split("--"))
+
+    return metadata
+
+
+def extract_basic_user_info(user_info: dict) -> dict:
+    """Extract a user's basic info from a Slack user object.
+
+    See https://api.slack.com/types/user for the structure of the user object.
+    """
+    return {
+        "id": user_info.get("id"),
+        "name": user_info.get("name"),
+        "is_bot": user_info.get("is_bot"),
+        "email": user_info.get("profile", {}).get("email"),
+        "display_name": user_info.get("profile", {}).get("display_name"),
+        "real_name": user_info.get("real_name"),
+        "timezone": user_info.get("tz"),
+    }
+
+
+def is_user_a_bot(user: dict) -> bool:
+    """Check if a Slack user object represents a bot.
+
+    Bots are users with the "is_bot" flag set to true.
+    USLACKBOT is the user object for the Slack bot itself and is a special case.
+
+    See https://api.slack.com/types/user for the structure of the user object.
+    """
+    return user.get("is_bot") or user.get("id") == "USLACKBOT"
+
+
+def is_user_deleted(user: dict) -> bool:
+    """Check if a Slack user object represents a deleted user.
+
+    See https://api.slack.com/types/user for the structure of the user object.
+    """
+    return user.get("deleted", False)
+
+
+async def async_paginate(
+    func: Callable,
+    response_key: Optional[str] = None,
+    limit: Optional[int] = MAX_PAGINATION_LIMIT,
+    next_cursor: Optional[str] = None,
+    *args,
+    **kwargs,
+) -> tuple[list, Optional[str]]:
+    """Paginate a Slack AsyncWebClient's function results.
+
+    The purpose is to abstract the pagination work and make it easier for the LLM to retrieve the
+    amount of items requested by the user, regardless of limits imposed by the Slack API. We still
+    return the next cursor, if needed to paginate further.
+    """
+    results = []
+    while len(results) < limit:
```

Comment:
what if results is longer than the max pagination limit?

Reply:
It's not going to be.

The `slack_limit` variable makes the Slack API return no more than `min(limit - len(results), MAX_PAGINATION_LAIMIT)`.

Then, it extend results with the items returned and continue looping if it hasn't reached the limit, and slack indicates there's more to retrieve.

Reply:
Good answer. Write this as comments in the code